pkg/shoot_client/tunnel/tunnel.go

```diff
@@ -0,0 +1,221 @@
+	remoteBond          net.IP
+	lastSeen            time.Time
+	creationComplete    bool
+	lastCreationFailed  *time.Time
+	creationFailedCount int
+	lastError           error
+}
+
+func (d *kubeApiserverData) setLastSeen() {
+	d.lock.Lock()
+	defer d.lock.Unlock()
+	d.lastSeen = time.Now()
+}
+
+func (d *kubeApiserverData) needsUpdate(tunnelConfig IP6Tunnel) bool {
+	d.lock.Lock()
+	defer d.lock.Unlock()
+
+	if d.podIP != tunnelConfig.KubeAPIServerPodIP {
+		return true
+	}
+	if d.creationComplete {
+		return false
+	} else if d.lastCreationFailed != nil {
+		return time.Since(*d.lastCreationFailed) > 30*time.Second
+	}
+	return true
+}
+
+func (d *kubeApiserverData) update() {
+	d.lock.Lock()
+	defer d.lock.Unlock()
+
+	name := fmt.Sprintf("bond0-ip6tnl-%02x", d.remoteBond[len(d.remoteBond)-1])
+
+	if err := network.DeleteLinkByName(name); err != nil {
+		d._setFailed(fmt.Errorf("failed to delete link %s: %w", name, err))
+		return
+	}
+
+	if err := network.CreateTunnelIP6Tnl(name, d.localBond, d.remoteBond); err != nil {
+		d._setFailed(fmt.Errorf("failed to create tunnel %s: %w", name, err))
+		return
+	}
+	d.log.Info("tunnel created", "name", name)
+
+	link, err := netlink.LinkByName(name)
+	if err != nil {
+		d._setFailed(fmt.Errorf("failed to get link %s: %w", name, err))
+		return
+	}
+
+	ip := net.ParseIP(d.podIP)
+	if ipv4 := ip.To4(); ipv4 != nil {
+		ip = ipv4
+	}
+	if ip == nil {
+		d._setFailed(fmt.Errorf("failed to parse pod IP %s: %w", d.podIP, err))
+		return
+	}
+
+	if err := network.RouteReplace(d.log, &net.IPNet{IP: ip, Mask: net.CIDRMask(len(ip)*8, len(ip)*8)}, link); err != nil {
+		d._setFailed(fmt.Errorf("failed to replace route %s: %w", name, err))
+		return
+	}
+
+	d.creationComplete = true
+	d.lastError = nil
+	d.lastCreationFailed = nil
+	d.creationFailedCount = 0
+}
+
+func (d *kubeApiserverData) delete() error {
+	d.lock.Lock()
+	defer d.lock.Unlock()
+
+	name := fmt.Sprintf("bond0-ip6tnl-%02x", d.remoteBond[len(d.remoteBond)-1])
+	return network.DeleteLinkByName(name)
+}
+
+func (d *kubeApiserverData) _setFailed(err error) {
+	d.lastCreationFailed = ptr.To(time.Now())
+	d.creationFailedCount++
+	d.lastError = err
+	d.log.Error(err, "failed to update tunnel controller")
+}
+
+func (d *kubeApiserverData) isOutdated() bool {
+	d.lock.Lock()
+	defer d.lock.Unlock()
+	return time.Since(d.lastSeen) > 10*time.Minute
+}
+
+// Controller is a server receiving UDP requests to create ipv6tnl devices.
+type Controller struct {
+	kubeApiservers map[string]*kubeApiserverData
+	nextClean      time.Time
+}
+
+// Run runs the tunnel controller
+func (c *Controller) Run(log logr.Logger) error {
+	ips, err := network.GetLinkIPAddressesByName("bond0", network.ScopeUniverse)
+	if err != nil {
+		return err
+	}
+	if len(ips) == 0 {
+		return fmt.Errorf("no IP addresses found for bond0")
+	}
+	if len(ips[0]) != 16 {
+		return fmt.Errorf("expected ipv6 address for bond0, got %s", ips[0])
+	}
+
+	localBond := ips[0]
```

Comment:
The bond device is created and configured by our code to have a single IP. It can only happen if someone tampers with it. If the tunnel controller listen on a different IP, it will not receive any requests from the kube-apiserver pods, as the IP is known and fixed there.
Independent of the IP address someone could send own requests to the tunnel controller, but what can you do with that as the firewall rules only allows `RELATED,ESTABLISHED` packets?